File: Sequence/q100-q199/q110-109/q110/q110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (root == NULL) {
            return true;
        }
        bool balanced = true;
        height(root, & balanced);
        return balanced;
    }
    
    int height(TreeNode* root, bool* balanced) {
        if (root == NULL) {
            return 0;
        }
        int left_height = height(root->left, balanced);
        int right_height = height(root->right, balanced);
        if (abs(left_height - right_height) > 1) {
            *balanced = false;
            return 1;
        }
        return 1 + max(left_height, right_height);
    }
};

File: Sequence/q100-q199/q110-109/q110/q110.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        def height(root: TreeNode) -> int:
            nonlocal balanced
            if root == None:
                return 0
            left_height = height(root.left)
            right_height = height(root.right)
            if abs(left_height - right_height) > 1:
                balanced = False
            return 1 + max(left_height, right_height)
        balanced = True
        height(root)
        return balanced
